danceradvent/public/articles/2015/7-medium-scale-part-5.pod

```diff
@@ -0,0 +1,208 @@
+spec does not allow browsers to assume that identical C<POST> calls are
+idempotent, however, so we use that verb for CRUD's create operation,
+leaving C<PUT> = update.
+
+That's enough design philosophy; how does all that affect our Dancer
+code? We might decide that since we've almost got a REST API here that
+we might want to push it the rest of the way.
+
+Let's move the CRUD-like route handlers in C<lib/App/MajorFeature.pm> to
+C<lib/App/API/SubFeature.pm>, and also refine the programming interface
+a bit:
+
+ prefix '/api' => sub {
+     prefix '/subfeature' => sub {
+         post '/'    => \&_post;
+         get  '/'    => \&_get;
+         put  '/:id' => \&_put;
+         del  '/:id' => \&_del;
+     };
+ };
+
+My choice to use leading underscores in these module-internal route
+handler function names lets us parallel the Dancer DSL keyword naming
+scheme.
+
+You might instead prefer to make the HTTP-to-CRUD mapping more explicit:
+
+ prefix '/api' => sub {
+     prefix '/subfeature' => sub {
+         post '/'    => \&_create;
+         get  '/'    => \&_read;
+         put  '/:id' => \&_update;
+         del  '/:id' => \&_delete;
+     };
+ };
+
+As in the previous article in this series, the leading underscore
+convention saves us some aggravation here since C<delete> is a Perl
+keyword, and C<read> is a core Perl function.
+
+The only other thing to change here is that REST APIs normally return
+JSON or XML. I prefer JSON since that's directly usable by the
+JavaScript code that made the Ajax call to one of these APIs. One of the
+best features of Dancer is that automatic JSON encoding is built right
+in. We might define the C<_create()> function referenced above as:
+
+ sub _create {
+     my $params = request->params;
+     my $db     = get_db_conn;
+ 
+     if ($db->create_something($params->{arg1}, $params->{arg2})) {
+         return {
+             success => JSON::true,
+             id      => $db->last_insert_id(),
+         };
+     }
+     else {
+         return {
+             success => JSON::false,
+             error   => 'failed to create record: ' . $db->last_error(),
+         };
+     }
+ }
+
+In Dancer, a hash reference returned from a route handler is
+automatically JSON-encoded, and the HTTP response's content type is set
+to C<application/json>. Parsing and using that reply on the client side
+in JavaScript is therefore trivial.
+
+If your new REST API route handlers previously returned HTML, and you
+have existing code that still needs this data in HTML form, I prefer to
+separate those Ajax call handlers into a different section of the URL
+hierarchy: C</parts>. The rule is simple: C</api> routes return JSON,
+whereas C</parts> routes return HTML fragments. Those fragments are
+intended to be inserted into an existing DOM via JavaScript. The
+C</parts> section of the URL hierarchy separates out such route handlers
+from those that return whole HTML pages.
```

Comment:
This is where separating to different apps works in your favor. If one app sets a JSON serializer and the other one does not, one can serve JSON and the other HTML.